### ISSUE TITLE
Changefeed "subscriber groups"

### ISSUE THREAD

**Issue by nickcarenza** [rethinkdb/rethinkdb#5038]
I would like to be able to distribute the handling of changes by specifying a named group when subscribing to a changefeed and ask that rethink only return a change entry to a single member of the group.

```
        cursor, err = r.Table("things").Filter(
                map[string]interface{}{
                        "old_val": nil,
                },
        ).Changes(r.ChangesOpts{
                Squash: true,
                Group: "task-handler",
        }).Run(rethink)
```

I would still be responsible for making sure that each thing was processed correctly.

**Comment by coffeemug** [comment]
A workaround for now would be to use this pattern/library: http://rethinkdb.com/docs/publish-subscribe/javascript/.

I agree this would be very cool to support natively. @nickcarenza -- could you give a few more details on your use case? (understanding use cases always helps design the feature better)

**Comment by ChrisMcKenzie** [comment]
:+1:

**Comment by nickcarenza** [comment]
Sure thing and thanks for the quick response!

First of all, my reasons for using rethink are ReQL, the admin ui and easy clustering.

I have a distributed group of workers processing messages coming from another pub sub system and I am storing the messages i was unable to process for whatever reason in rethink.

```
{
    "Created": Wed Nov 04 2015 11:18:45 GMT+00:00 ,
    "Errors": [
        "Type: required, Description: brand_slug is required, Field: brand_slug"
    ] ,
    "Message": "{...}" ,
    "Process":  "event-writer" ,
    "Reason":  "Invalid event" ,
    "Subject":  "session:worker" ,
    "id":  "010a2b20-a7d9-490a-938d-d63effa7764d"
}
```

I manually determine what's wrong with them and use the data explorer to manipulate the Message and write it to a new table with `r.json`.

```
# Move objectified json message to staging -- with added field
r.db('deadletter').table('messages').filter({
    Subject:"session:worker",
    Process:"event-writer",
    Errors:["Type: required, Description: brand_slug is required, Field: brand_slug"]
}).forEach(function(row){
  return r.db('deadletter').table('staging').insert(
    r.json(row('Message'))
      .merge({brand_slug:'BrandSlug'})
    );
})
```

At first I was going to make a script to process them from there but then I thought I would try the changes api. I just need each cleaned message to be processed only once. To accomplish this, I am attempting to update the document with some meta data and checking for that in the other worker processes. With squash, some processes shouldn't even have to attempt the update so that is nice.

**Comment by coffeemug** [comment]
I see -- thank you!

**Comment by danielmewes** [comment]
This is very similar to https://github.com/rethinkdb/rethinkdb/issues/3184 . I think I'm going to close the other one, since this one has a particular API proposal in it.

**Comment by marshall007** [comment]
Related to #4820 since the server would need some mechanism of uniquely identifying clients in order to make this possible. Also possibly #4133 depending on the use-case.

**Comment by jdrago999** [comment]
Hi - just curious if any progress has already been made on this (message is only processed once)?

The existing feature of "broadcast changes to all subscribers simultaneously" is great, but having the option to only send a change to one connected subscriber would really be awesome.

**Comment by coffeemug** [comment]
We haven't made progress on the issue. The good news is that this isn't very hard to do; the bad news is that we have a really deep pipeline of features that's quite challenging to prioritize. We'll try to get this scheduled ASAP, I'll talk to @danielmewes about our release plans, we'll possibly reshuffle a few things, and report back.

**Comment by nickcarenza** [comment]
So you think this is something a new contributor could reasonably accomplish?

**Comment by coffeemug** [comment]
TBH, I don't think so (although a sufficiently driven, smart person could *probably* get it done). There is an enormous amount of knowhow locked up in the codebase and the minds of experienced RethinkDB developers. If you have the knowhow, implementing this is pretty easy; if you don't, it's impossibly hard.

One of our goals for 2016 is to make the db server codebase more accessible (via documenting internals, writing contributor tutorials, etc.) but unfortunately for now it feels a little bit like voodoo magic to new contributors.

**Comment by geoah** [comment]
Is this currently in the roadmap for one of the next releases?

**Comment by danielmewes** [comment]
@geoah Not sure which release this is going to be in, but we'll try our best. It's a feature that has been asked for a lot, so I think chances are good that it will make it into one of the next few releases.

**Comment by mans2singh** [comment]
+1

**Comment by johanneslumpe** [comment]
@danielmewes If groups will enable us to only receive the change on one of the connected nodes, wouldn't this be a very brittle thing, given the fact that there is now ACK sent? Do you think it would become a requirement to make sure that at least one group member definitely received the message?

**Comment by danielmewes** [comment]
@johanneslumpe Well in general there are the two options of "at most once" and "at least once" delivery. Without an ACK, this would provide "at most once" delivery, and indeed messages could get lost if a client fails in the wrong moment as you say.
We could add some form of client ack to allow for "at least once" delivery, though it's significantly more complicated. For example there's the question of how long the server should wait for a client to acknowledge a message before it assumes that the message got lost etc,

**Comment by johanneslumpe** [comment]
@danielmewes I can see that "at least once" would be more complex to implement. Figuring out how long to wait is definitely something to think about, but without this type of validation, change feed groups feel a bit like playing Russian roulette. And that would be especially dangerous if one wanted to build an application which require change feeds at their core, for example the real time market place: if we wanted to push out updates for an item's price to our users and we're using a 3rd party pubsub provider, then we definitely don't want to send out the message twice, because that might cost us quite a bit if there are a lot of connected users. Another problematic thing would be if we need to send out emails when a certain prop for a document changed - those emails shouldn't be sent out twice.

**Comment by cur3n4** [comment]
Hello, is this in for any near term release? Anyway I could help with this?

**Comment by esatterwhite** [comment]
This is something I'd also love to see. Even if rethink just did a round robin between connected clients of the same change feed name / query.

The practical use of blanket server wide broadcasting is small. I think the idea of a `worker` pool is much more realistic.

**Comment by jdrago999** [comment]
@esatterwhite I agree but in the case of distributed RethinkDB servers, they'll also need to keep track of which messages have been delivered + acked/unacked.

I also would love to see this functionality.

**Comment by esatterwhite** [comment]
@jdrago999 I don't think managing message acknowledgment is something rethink should have to do. It's a change feed. missing a message isn't really a bad thing. You'll get another one. Also, the client is always connected to the data. you can manage that with data and the change feed query.

I think all rethink should be able to do is group connections and cycle between them in some fashion. I like the firehose / fire and forget nature of changefeeds, I just want to be able to aim it a little better.

**Comment by cedbale** [comment]
any visibility about the possibility to include this feature into the roadmap after https://github.com/rethinkdb/rethinkdb/issues/6747 announcement ? 😊
